### ISSUE TITLE
Bug: [no-unreachable] doesn't treat `import`s as hoisted

### ISSUE THREAD

**Issue by kirkwaiblinger** [eslint/eslint#21273]
### Environment

ESLint Playground

### What parser are you using?

Default (Espree)

### Playground Code
```js
/*eslint no-unreachable: "error"*/

import foo from 'foo';

console.log('bar');

throw new Error();

import baz from 'baz';
```

### ESLint Configuration 
<details> 
 <summary> Tap to see full config </summary> 

```js
import { defineConfig } from "eslint/config";

export default defineConfig([
    {
        "languageOptions": {
            "sourceType": "module"
        }
    }
]);
```
 </details>

### ESLint Output
```
9:1 Unreachable code. (no-unreachable)
```

### What did you expect to happen?

no lint report

### What actually happened?

9:1 Unreachable code. (no-unreachable)

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW5yZWFjaGFibGU6IFwiZXJyb3JcIiovXG5cbmltcG9ydCBmb28gZnJvbSAnZm9vJztcblxuY29uc29sZS5sb2coJ2JhcicpO1xuXG50aHJvdyBuZXcgRXJyb3IoKTtcblxuaW1wb3J0IGJheiBmcm9tICdiYXonOyIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fSwic291cmNlVHlwZSI6Im1vZHVsZSJ9fX0=

### Participation

- [ ] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

Wasn't sure whether to report this as a bug against the code path analysis itself. The code explorer shows the import as unreachable due to interpreting it as ordinary source code, rather than being hoisted for the purposes of control flow ([link](https://explorer.eslint.org/#eslint-explorer=v2.eyJzdGF0ZSI6eyJ0b29sIjoicGF0aCIsImNvZGUiOnsiamF2YXNjcmlwdCI6ImltcG9ydCBmb28gZnJvbSAnZm9vJztcblxuY29uc29sZS5sb2coJ2JhcicpO1xuXG50aHJvdyBuZXcgRXJyb3IoKTtcblxuaW1wb3J0IGJheiBmcm9tICdiYXonOyIsImpzb24iOiIvKipcbiAqIFR5cGUgb3IgcGFzdGUgc29tZSBKU09OIGhlcmUgdG8gbGVhcm4gbW9yZSBhYm91dFxuICogdGhlIHN0YXRpYyBhbmFseXNpcyB0aGF0IEVTTGludCBjYW4gZG8gZm9yIHlvdS5cbiAqXG4gKiBUaGUgdGFicyBhcmU6XG4gKlxuICogLSBBU1QgLSBUaGUgQWJzdHJhY3QgU3ludGF4IFRyZWUgb2YgdGhlIGNvZGUsIHdoaWNoIGNhblxuICogICBiZSB1c2VmdWwgdG8gdW5kZXJzdGFuZCB0aGUgc3RydWN0dXJlIG9mIHRoZSBjb2RlLiBZb3VcbiAqICAgY2FuIHZpZXcgdGhpcyBzdHJ1Y3R1cmUgYXMgSlNPTiBvciBpbiBhIHRyZWUgZm9ybWF0LlxuICpcbiAqIFlvdSBjYW4gY2hhbmdlIHRoZSB3YXkgdGhhdCB0aGUgSlNPTiBjb2RlIGlzIGludGVycHJldGVkXG4gKiBieSBjbGlja2luZyBcIkpTT05cIiBpbiB0aGUgaGVhZGVyIGFuZCBzZWxlY3RpbmcgZGlmZmVyZW50XG4gKiBvcHRpb25zLlxuICpcbiAqIFRoaXMgZXhhbXBsZSBpcyBpbiBKU09OQyBtb2RlLCB3aGljaCBhbGxvd3MgY29tbWVudHMuXG4gKi9cblxue1xuICAgIFwia2V5MVwiOiBbdHJ1ZSwgZmFsc2UsIG51bGxdLFxuICAgIFwia2V5MlwiOiB7XG4gICAgICAgIFwia2V5M1wiOiBbMSwgMiwgXCIzXCIsIDFlMTAsIDFlLTNdXG4gICAgfVxufSIsIm1hcmtkb3duIjoiPCEtLVxuVHlwZSBvciBwYXN0ZSBzb21lIE1hcmtkb3duIGhlcmUgdG8gbGVhcm4gbW9yZSBhYm91dFxudGhlIHN0YXRpYyBhbmFseXNpcyB0aGF0IEVTTGludCBjYW4gZG8gZm9yIHlvdS5cblxuVGhlIHRhYnMgYXJlOlxuXG4tIEFTVCAtIFRoZSBBYnN0cmFjdCBTeW50YXggVHJlZSBvZiB0aGUgY29kZSwgd2hpY2ggY2FuXG5iZSB1c2VmdWwgdG8gdW5kZXJzdGFuZCB0aGUgc3RydWN0dXJlIG9mIHRoZSBjb2RlLiBZb3VcbmNhbiB2aWV3IHRoaXMgc3RydWN0dXJlIGFzIEpTT04gb3IgaW4gYSB0cmVlIGZvcm1hdC5cblxuWW91IGNhbiBjaGFuZ2UgdGhlIHdheSB0aGF0IHRoZSBNYXJrZG93biBjb2RlIGlzIGludGVycHJldGVkXG5ieSBjbGlja2luZyBcIk1hcmtkb3duXCIgaW4gdGhlIGhlYWRlciBhbmQgc2VsZWN0aW5nIGRpZmZlcmVudFxub3B0aW9ucy5cblxuVGhpcyBleGFtcGxlIGlzIGluIENvbW1vbk1hcmsgbW9kZS5cbi0tPlxuXG4jIEVTTGludCBNYXJrZG93biBFeGFtcGxlXG5cblRoaXMgaXMgYW4gZXhhbXBsZSBvZiBhIE1hcmtkb3duIGZpbGUgdGhhdCBjYW4gYmUgcGFyc2VkXG5ieSBFU0xpbnQuIE1hcmtkb3duIGlzIGEgc2ltcGxlIG1hcmt1cCBsYW5ndWFnZSB0aGF0IGlzXG5vZnRlbiB1c2VkIGZvciBkb2N1bWVudGF0aW9uLlxuXG4jIyBGZWF0dXJlc1xuXG4tIE1ha2UgdGhpbmdzICppdGFsaWMqLCAqKmJvbGQqKiwgb3IgYGNvZGVgXG4tIENyZWF0ZSBbbGlua3NdKGh0dHBzOi8vZXNsaW50Lm9yZylcbi0gU3VwcG9ydHMgSFRNTCA8c3BhbiBzdHlsZT1cImNvbG9yOiByZWQ7XCI-ZWxlbWVudHM8L3NwYW4-XG4tIExpc3RzXG4gIC0gTmVzdGVkIGxpc3RzIiwiY3NzIjoiLyoqXG4gKiBUeXBlIG9yIHBhc3RlIHNvbWUgQ1NTIGhlcmUgdG8gbGVhcm4gbW9yZSBhYm91dFxuICogdGhlIHN0YXRpYyBhbmFseXNpcyB0aGF0IEVTTGludCBjYW4gZG8gZm9yIHlvdS5cbiAqXG4gKiBUaGUgdGFicyBhcmU6XG4gKlxuICogLSBBU1QgLSBUaGUgQWJzdHJhY3QgU3ludGF4IFRyZWUgb2YgdGhlIGNvZGUsIHdoaWNoIGNhblxuICogICBiZSB1c2VmdWwgdG8gdW5kZXJzdGFuZCB0aGUgc3RydWN0dXJlIG9mIHRoZSBjb2RlLiBZb3VcbiAqICAgY2FuIHZpZXcgdGhpcyBzdHJ1Y3R1cmUgYXMgSlNPTiBvciBpbiBhIHRyZWUgZm9ybWF0LlxuICpcbiAqIFlvdSBjYW4gY2hhbmdlIHRoZSB3YXkgdGhhdCB0aGUgQ1NTIGNvZGUgaXMgaW50ZXJwcmV0ZWRcbiAqIGJ5IGNsaWNraW5nIFwiQ1NTXCIgaW4gdGhlIGhlYWRlciBhbmQgc2VsZWN0aW5nIGRpZmZlcmVudFxuICogb3B0aW9ucy5cbiAqL1xuXG5AaW1wb3J0IHVybCgnaHR0cHM6Ly9mb250cy5nb29nbGVhcGlzLmNvbS9jc3MyP2ZhbWlseT1Sb2JvdG86d2dodEA0MDA7NzAwJmRpc3BsYXk9c3dhcCcpO1xuXG5ib2R5IHtcblx0Zm9udC1mYW1pbHk6IHNhbnMtc2VyaWY7XG59XG5cbmgxIHtcblx0Y29sb3I6ICMzMzM7XG59XG5cbnAge1xuXHRtYXJnaW46IDFlbSAwO1xufSIsImh0bWwiOiI8IURPQ1RZUEUgaHRtbD5cbjwhLS1cblR5cGUgb3IgcGFzdGUgc29tZSBIVE1MIGhlcmUgdG8gbGVhcm4gbW9yZSBhYm91dFxudGhlIHN0YXRpYyBhbmFseXNpcyB0aGF0IEVTTGludCBjYW4gZG8gZm9yIHlvdS5cblxuVGhlIHRhYnMgYXJlOlxuXG4tIEFTVCAtIFRoZSBBYnN0cmFjdCBTeW50YXggVHJlZSBvZiB0aGUgY29kZSwgd2hpY2ggY2FuXG5iZSB1c2VmdWwgdG8gdW5kZXJzdGFuZCB0aGUgc3RydWN0dXJlIG9mIHRoZSBjb2RlLiBZb3VcbmNhbiB2aWV3IHRoaXMgc3RydWN0dXJlIGFzIEpTT04gb3IgaW4gYSB0cmVlIGZvcm1hdC5cbi0tPlxuXG48aHRtbCBsYW5nPVwiZW5cIj5cbiAgICA8aGVhZD5cbiAgICAgICAgPG1ldGEgY2hhcnNldD1cIlVURi04XCI-XG4gICAgICAgIDx0aXRsZT5IVE1MPC90aXRsZT5cbiAgICA8L2hlYWQ-XG4gICAgPGJvZHk-XG4gICAgICAgIDxwPlRleHQ8L3A-XG4gICAgPC9ib2R5PlxuPC9odG1sPiJ9LCJsYW5ndWFnZSI6ImphdmFzY3JpcHQiLCJqc09wdGlvbnMiOnsicGFyc2VyIjoiZXNwcmVlIiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVzVmVyc2lvbiI6ImxhdGVzdCIsImlzSlNYIjp0cnVlfSwianNvbk9wdGlvbnMiOnsianNvbk1vZGUiOiJqc29uYyIsImFsbG93VHJhaWxpbmdDb21tYXMiOmZhbHNlfSwiY3NzT3B0aW9ucyI6eyJjc3NNb2RlIjoiY3NzIiwidG9sZXJhbnQiOmZhbHNlfSwibWFya2Rvd25PcHRpb25zIjp7Im1hcmtkb3duTW9kZSI6ImNvbW1vbm1hcmsiLCJtYXJrZG93bkZyb250bWF0dGVyIjoib2ZmIiwibWFya2Rvd25NYXRoIjpmYWxzZX0sImh0bWxPcHRpb25zIjp7InRlbXBsYXRlRW5naW5lU3ludGF4Ijoibm9uZSIsImZyb250bWF0dGVyIjpmYWxzZX0sIndyYXAiOmZhbHNlLCJ2aWV3TW9kZXMiOnsiYXN0VmlldyI6InRyZWUiLCJzY29wZVZpZXciOiJmbGF0IiwicGF0aFZpZXciOiJncmFwaCJ9LCJwYXRoSW5kZXgiOnsiaW5kZXgiOjAsImluZGV4ZXMiOjF9LCJlc3F1ZXJ5U2VsZWN0b3IiOnsic2VsZWN0b3IiOiIifX0sInZlcnNpb24iOjB9)). This seems kind of wrong to me? Imports really aren't part of the top-level execution code path at all; they're more like their own thing that happens before the top-level execution.

**Comment by tahakotil** [comment]
Reproduced on ESLint v10.9.1 with the issue's minimal module: line 9 is still reported as `Unreachable code`. The import is instantiated before module evaluation, so treating the declaration as a post-throw runtime statement appears to be a false positive.

I can prepare a focused `no-unreachable` fix plus regression coverage once the issue is accepted.

AI disclosure: I used an AI coding assistant to help investigate and draft this note; the reproduction command and output were checked locally.

**Comment by crimsonjay0** [comment]
I also think this is a bug. I'd like to be assigned to this.

**Comment by mdjermanovic** [comment]
I agree this is a bug. Marking as accepted, PR is welcome. Assigning to @tahakotil as asked first.
